src/SonarScanner.MSBuild.Common/ConsoleLogger.cs

```diff
@@ -137,6 +138,38 @@ public void WriteUIWarnings(string outputFolder)
         }
     }
 
+    /// <summary>
+    /// Saves a telemetry message for later processing.
+    /// The <paramref name="value"/> parameter must be a primitive JSON type, like a number, a String, or a Boolean.
+    /// </summary>
+    public void AddTelemetryMessage(string key, object value) =>
+        telemetryMessages.Add(new(key, value));
+
+    public void WriteTelemetry(string outputFolder)
+    {
+        var telemetryMessagesJson = new StringBuilder();
+        foreach (var message in telemetryMessages)
+        {
+            telemetryMessagesJson.AppendLine(ParseMessage(message));
+        }
+
+        var path = Path.Combine(outputFolder, FileConstants.TelemetryFileName);
+        var telemetry = telemetryMessagesJson.ToString();
+        fileWrapper.AppendAllText(path, telemetry);
```

Comment:
This may fail. Add a try/catch(IOException) block and log the error. We do not want to fail the begin or end step because we could not write to a telemetry file.

Reply:
🤦  indeed.